hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/scm/proxy/SCMSecurityProtocolFailoverProxyProvider.java

```diff
@@ -0,0 +1,279 @@
+    if (LOG.isDebugEnabled()) {
+      int currentIndex = getCurrentProxyIndex();
+      LOG.debug("Failing over SCM Security proxy to index: {}, nodeId: {}",
+          currentIndex, scmNodeIds.get(currentIndex));
+    }
+  }
+
+  /**
+   * Performs fail-over to the next proxy.
+   */
+  public void performFailoverToNextProxy() {
+    int newProxyIndex = incrementProxyIndex();
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("Incrementing SCM Security proxy index to {}, nodeId: {}",
+          newProxyIndex, scmNodeIds.get(newProxyIndex));
+    }
+  }
+
+  /**
+   * Update the proxy index to the next proxy in the list.
+   * @return the new proxy index
+   */
+  private synchronized int incrementProxyIndex() {
+    currentProxyIndex = (currentProxyIndex + 1) % scmProxies.size();
+    currentProxySCMNodeId = scmNodeIds.get(currentProxyIndex);
+    return currentProxyIndex;
+  }
+
+  public RetryPolicy getRetryPolicy() {
+    // Client will attempt up to maxFailovers number of failovers between
+    // available SCMs before throwing exception.
+    RetryPolicy retryPolicy = new RetryPolicy() {
+      @Override
+      public RetryAction shouldRetry(Exception exception, int retries,
+          int failovers, boolean isIdempotentOrAtMostOnce)
+          throws Exception {
+
+        if (LOG.isDebugEnabled()) {
+          if (exception.getCause() != null) {
+            LOG.debug("RetryProxy: SCM Security Server {}: {}: {}",
+                getCurrentProxySCMNodeId(),
+                exception.getCause().getClass().getSimpleName(),
+                exception.getCause().getMessage());
+          } else {
+            LOG.debug("RetryProxy: OM {}: {}", getCurrentProxySCMNodeId(),
+                exception.getMessage());
+          }
+        }
+
+        // For AccessControl Exception where Client is not authentica
+        if (HAUtils.isAccessControlException(exception)) {
+          return RetryAction.FAIL;
+        }
+
+        // Perform fail over to next proxy, as right now we don't have any
+        // suggested leader ID from server, we fail over to next one.
+        // TODO: Act based on server response if leader id is passed.
+        performFailoverToNextProxy();
+        return getRetryAction(FAILOVER_AND_RETRY, failovers);
+      }
+
+      private RetryAction getRetryAction(RetryDecision fallbackAction,
+          int failovers) {
+        if (failovers < maxRetryCount) {
+          return new RetryAction(fallbackAction, getRetryInterval());
+        } else {
+          return RetryAction.FAIL;
+        }
+      }
+    };
+
+    return retryPolicy;
+  }
+
+
+  @Override
+  public Class< SCMSecurityProtocolPB > getInterface() {
+    return SCMSecurityProtocolPB.class;
+  }
+
+  @Override
+  public void close() throws IOException {
+    for (ProxyInfo<SCMSecurityProtocolPB> proxyInfo : scmProxies.values()) {
+      if (proxyInfo != null) {
+        RPC.stopProxy(proxyInfo.proxy);
```

Comment:
can we go through the proxy map, call stopProxy and null the value in the map?

Reply:
Done